### DISCUSSION TITLE
Clearing Uninstalled Packages from Cache

### DISCUSSION THREAD

**Discussion by jdholtz** [aokellermann/yaycache#12] [Q&A] [answered]
Hello. Just started using this tool and it's very helpful. I was wondering if there is an option to remove all AUR packages not currently installed on my system. This would be what the question `Do you want to remove all other AUR packages from cache?` when running `yay -Sca` does.

For example, I download an AUR package. Then I remove it. However, the directory for the AUR package still exists in the cache. Is there an option to remove these directories? If not, I think this would need to be implemented a bit differently than paccache, as the yay cache stores each package inside its own directory (and I don't think paccache has the logic for handling that).

The biggest reason I’d want this is to use it within a libalpm hook. Another possible solution is also running yay -Sca in a hook and only saying yes to the first question.

**Comment by aokellermann** [comment, marked as answer]
Yes

```sh
$ yaycache --help | grep uninstalled
  -u, --uninstalled       	target uninstalled packages
```

**Reply by jdholtz** [reply]
Thanks. However, this doesn't remove the package's directory in the cache. For example, I install an AUR package with `yay -S <pkg>`. I can see the folder exists at `~/.cache/yay/<pkg>`. Then, I uninstall it with `yay -Rns <pkg>` and still see the folder `~/.cache/yay/<pkg>` exists.

When I run `yaycache -ru`, it reports back `no candidate packages found for pruning` and the folder `~/.cache/yay/<pkg>` still exists. My question is if, in its current state, I can remove the entire package's directory (`~/.cache/yay/<pkg>`), which is what the first question of `yay -Sca` does.

**Reply by aokellermann** [reply]
If you don't specify `-k`, `yaycache` (like `paccache`) will default to keeping 3 latest versions of the package. Pass `-k0` to keep none. 
- `yaycache -ruk0`: removes all versions of `.pkg` of uninstalled aur packages
- `yaycache -ruk0 --remove-build-files`: removes all versions of `.pkg` of uninstalled AUR packages and any files that untracked files `makepkg` used to build the package (AUR packages are just git repos containing a `PKGBUILD` and a few other small files typically). 

There is no way in current state to remove `~/.cache/yay/<pkg>` but using the above commands the bulky files will be removed.

**Reply by jdholtz** [reply]
> There is no way in current state to remove ~/.cache/yay/<pkg> but using the above commands the bulky files will be removed

Would you want me to submit a PR to add support for this when the -u flag is specified?

**Reply by aokellermann** [reply]
Can you give some context as to why you want to fully delete these directories? I doubt they are taking up much space. 

I would just go with the `yay -Sca` solution you mentioned or do it in pure bash:

```bash
cd .cache/yay
comm -23 <(find . -maxdepth 1 -mindepth 1 -type d | sort | cut -f2 -d'/') <(yay -Qqm | sort) | xargs rm -rf
```

Doing it in `yaycache` would probably require a new CLI argument which I'm apprehensive about.

**Reply by jdholtz** [reply]
> Can you give some context as to why you want to fully delete these directories? I doubt they are taking up much space.

Just to remove clutter on my filesystem and have a similar functionality in yaycache that yay -Sca has. However, since I can get the same functionality with the first part of `yay -Sca`, then I'll stick with that. Thanks!
